packages/react/dialog/src/Dialog.tsx

```diff
@@ -354,39 +370,44 @@ const DialogContentImpl = React.forwardRef<DialogContentImplElement, DialogConte
       ...contentProps
     } = props;
     const context = useDialogContext(CONTENT_NAME, __scopeDialog);
+    const portalContext = useDialogPortalContext(OVERLAY_NAME, __scopeDialog);
     const contentRef = React.useRef<HTMLDivElement>(null);
     const composedRefs = useComposedRefs(forwardedRef, contentRef);
+    const PortalWrapper = portalContext.isInsidePortal ? UnstablePortal : React.Fragment;
 
     // Make sure the whole tree has focus guards as our `Dialog` will be
     // the last element in the DOM (beacuse of the `Portal`)
     useFocusGuards();
 
     return (
-      <>
-        <FocusScope
-          asChild
-          loop
-          trapped={trapFocus}
-          onMountAutoFocus={onOpenAutoFocus}
-          onUnmountAutoFocus={onCloseAutoFocus}
-        >
-          <DismissableLayer
-            role="dialog"
-            id={context.contentId}
-            aria-describedby={ariaDescribedBy || context.descriptionId}
-            // If `aria-label` is set, ensure `aria-labelledby` is undefined as to avoid confusion.
-            // Otherwise fallback to an explicit `aria-labelledby` or the ID used in the
-            // `DialogTitle`
-            aria-labelledby={ariaLabel ? undefined : ariaLabelledBy || context.titleId}
-            aria-label={ariaLabel || undefined}
-            data-state={getState(context.open)}
-            {...contentProps}
-            ref={composedRefs}
-            onDismiss={() => context.onOpenChange(false)}
-          />
-        </FocusScope>
-        {process.env.NODE_ENV === 'development' && <LabelWarning contentRef={contentRef} />}
-      </>
+      <PortalWrapper {...(portalContext.isInsidePortal ? portalContext.props : undefined)}>
+        {/* Prevent nested `Dialog`s from portalling unless they use their own `Dialog.Portal` */}
```

Comment:
Can you elaborate on this a bit? I'm not sure I follow.

Reply:
```jsx
<Dialog.Portal>
  <Dialog.Content>

    <Dialog.Root>
      // this one shouldn't portal and it won't be cause I've provided 
      // `isInsidePortal=false` from `Dialog.Content` otherwise it would 
      // have inherited the portal wrapped around `Dialog.Content`.
      <Dialog.Content /> 
    </Dialog.Root>

    <Dialog.Root>
      <Dialog.Portal><Dialog.Content /></Dialog.Portal> // this one should
    </Dialog.Root>

  </Dialog.Content>
</Dialog.Portal>
```